In [1]:
import numpy as np
import pandas as pd

In [2]:
# importing and storing the data 

food_data = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', error_bad_lines = True)

In [3]:
# reading the data

food_data

Area Code      Area  Item Code           Item  Element Code  \
0              4   Algeria       2501     Population           511   
1              4   Algeria       2501     Population          5301   
2              4   Algeria       2901    Grand Total           664   
3              4   Algeria       2901    Grand Total           674   
4              4   Algeria       2901    Grand Total           684   
...          ...       ...        ...            ...           ...   
60938        181  Zimbabwe       2899  Miscellaneous          5142   
60939        181  Zimbabwe       2899  Miscellaneous           645   
60940        181  Zimbabwe       2899  Miscellaneous           664   
60941        181  Zimbabwe       2899  Miscellaneous           674   
60942        181  Zimbabwe       2899  Miscellaneous           684   

                                      Element             Unit     Y2014  \
0               Total Population - Both sexes     1000 persons  38924.00   
1                    Domestic supply quantity      1000 tonnes      0.00   
2               Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3      Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4          Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                       ...              ...       ...   
60938                                    Food      1000 tonnes     42.00   
60939     Food supply quantity (kg/capita/yr)               kg      3.06   
60940           Food supply (kcal/capita/day)  kcal/capita/day      3.00   
60941  Protein supply quantity (g/capita/day)     g/capita/day      0.10   
60942      Fat supply quantity (g/capita/day)     g/capita/day      0.04   

          Y2015     Y2016     Y2017     Y2018  
0      39728.00  40551.00  41389.00  42228.00  
1          0.00      0.00      0.00      0.00  
2       3379.00   3372.00   3341.00   3322.00  
3         94.35     94.72     92.82     91.83  
4         79.36     77.40     80.19     77.28  
...         ...       ...       ...       ...  
60938     46.00     33.00     19.00     16.00  
60939      3.33      2.35      1.33      1.08  
60940      4.00      3.00      1.00      1.00  
60941      0.11      0.08      0.04      0.04  
60942      0.05      0.03      0.02      0.01  

[60943 rows x 12 columns]

# Question 11

    What is the total sum of Animal Fat produced in 2014 and 2017 respectively?

In [4]:
# Grouping by Item

food_data.groupby('Item').sum()

Area Code  Item Code  Element Code      Y2014      Y2015  \
Item                                                                            
Alcohol, Non-Food        38447     752497       1515090    2403.00    2180.00   
Alcoholic Beverages      75016    1649136       2171381  102410.11   98783.72   
Animal Products          18060     397035         90990   11935.65   11811.73   
Animal fats              85279    1882494       2567303  209460.54  200675.72   
Apples and products      69280    1352989       1920418    9499.23   10559.15   
...                        ...        ...           ...        ...        ...   
Vegetables, Other        77779    1518715       2275409  155038.96  158104.08   
Vegetal Products         18060     391905         90990  107145.19  107064.17   
Wheat and products       87073    1639683       2654934  232670.13  234710.51   
Wine                     66206    1319535       1818328    4497.36    4251.81   
Yams                     50798     970905       1525745  200396.96  203151.78   

                         Y2016      Y2017      Y2018  
Item                                                  
Alcohol, Non-Food      2223.00    2348.00    2293.00  
Alcoholic Beverages   96958.75   95581.06   97847.27  
Animal Products       11661.69   11547.65   11578.61  
Animal fats          183314.22  269617.53  269648.27  
Apples and products    9853.26   10198.90    9640.51  
...                        ...        ...        ...  
Vegetables, Other    156218.90  157752.59  163987.21  
Vegetal Products     106792.72  107655.20  107775.39  
Wheat and products   239377.07  240047.62  242645.19  
Wine                   3872.09    4178.02    4039.32  
Yams                 220626.20  229174.59  221272.09  

[119 rows x 8 columns]

    total sum of Animal Fat produced in 2014
           
           =209460.54

    total sum of Animal Fat produced in 2017
    
            =269617.53

# Question 12

    What is the mean and standard deviation across the whole dataset for the year 2015 to 3 decimal places?

In [5]:
# mean of the data for 2015

food_data.mean()

Area Code        134.265576
Item Code       2687.176706
Element Code    3814.856456
Y2014            134.196282
Y2015            135.235966
Y2016            136.555222
Y2017            140.917765
Y2018            143.758381
dtype: float64

    mean Y2015 = 135.236

In [6]:
# standard deviation

food_data.std()

Area Code         72.605709
Item Code        146.055739
Element Code    2212.007033
Y2014           1567.663696
Y2015           1603.403984
Y2016           1640.007194
Y2017           1671.862359
Y2018           1710.782658
dtype: float64

        standard deviation Y2015 = 1603.404

# Question 13

    What is the total number and percentage of missing data in 2016 to 2 decimal places?

In [7]:
# total number of missing data in 2016

missing_data = food_data['Y2016'].isnull().sum()
missing_data

1535

In [8]:
# Number data in 2016

len_data = len(food_data['Y2016'])
len_data

60943

In [9]:
# Percentage of missing data in 2016

percentage = (missing_data/len_data)*100
percentage

2.5187470259094566

        total number of missing data = 1535
        
        percentage of missing data = 2.52

# Question 14

        Which year had the highest correlation with ‘Element Code’?

In [10]:
elementcode_year = food_data[['Element Code', 'Y2014',
           'Y2015', 'Y2016', 'Y2017', 'Y2018']]
elementcode_year

Element Code     Y2014     Y2015     Y2016     Y2017     Y2018
0               511  38924.00  39728.00  40551.00  41389.00  42228.00
1              5301      0.00      0.00      0.00      0.00      0.00
2               664   3377.00   3379.00   3372.00   3341.00   3322.00
3               674     94.90     94.35     94.72     92.82     91.83
4               684     80.06     79.36     77.40     80.19     77.28
...             ...       ...       ...       ...       ...       ...
60938          5142     42.00     46.00     33.00     19.00     16.00
60939           645      3.06      3.33      2.35      1.33      1.08
60940           664      3.00      4.00      3.00      1.00      1.00
60941           674      0.10      0.11      0.08      0.04      0.04
60942           684      0.04      0.05      0.03      0.02      0.01

[60943 rows x 6 columns]

In [11]:
corr = elementcode_year.groupby('Element Code').corr(method='pearson')

corr

Y2014     Y2015     Y2016     Y2017     Y2018
Element Code                                                        
511          Y2014  1.000000  0.999988  0.999951  0.999885  0.999791
             Y2015  0.999988  1.000000  0.999987  0.999947  0.999878
             Y2016  0.999951  0.999987  1.000000  0.999986  0.999944
             Y2017  0.999885  0.999947  0.999986  1.000000  0.999986
             Y2018  0.999791  0.999878  0.999944  0.999986  1.000000
...                      ...       ...       ...       ...       ...
5911         Y2014  1.000000  0.976942  0.992694  0.941934  0.942024
             Y2015  0.976942  1.000000  0.965195  0.863665  0.863140
             Y2016  0.992694  0.965195  1.000000  0.963790  0.963378
             Y2017  0.941934  0.863665  0.963790  1.000000  0.999409
             Y2018  0.942024  0.863140  0.963378  0.999409  1.000000

[90 rows x 5 columns]

In [12]:
corr.max()

Y2014    1.0
Y2015    1.0
Y2016    1.0
Y2017    1.0
Y2018    1.0
dtype: float64

# Question 15

    What year has the highest sum of Import Quantity?

In [13]:
# Grouping by Element and sum

element_grouping = food_data.groupby('Element').sum()
element_grouping

Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                   708993   14197445      28068795   
Export Quantity                            599910   11840553      26026133   
Fat supply quantity (g/capita/day)         675050   13535000       3435732   
Feed                                       176272    3538507       7282199   
Food                                       663295   13285035      25406622   
Food supply (kcal/capita/day)              674057   13511060       3329296   
Food supply quantity (kg/capita/yr)        658446   13185401       3163725   
Import Quantity                            688174   13795966      28834929   
Losses                                     274353    5424803      10292107   
Other uses (non-food)                      235554    4729749       8926728   
Processing                                 271940    5350416      10313310   
Production                                 526751   10450053      21388191   
Protein supply quantity (g/capita/day)     675050   13535000       3385502   
Residuals                                  623271   12421089      24066350   
Seed                                       103537    2035933       4211574   
Stock Variation                            571566   11329527      21464704   
Total Population - Both sexes                6020     112545         22995   
Tourist consumption                         50308    1486528       2869905   

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein 

In [14]:
# printing the row with Import Quantity 

element_grouping.loc['Import Quantity']

Area Code         688174.00
Item Code       13795966.00
Element Code    28834929.00
Y2014             274144.48
Y2015             267018.46
Y2016             286582.78
Y2017             294559.09
Y2018             287997.09
Name: Import Quantity, dtype: float64

     the year with the highest sum of Import Quantity 
           
           = 2017

# Question 16

    What is the total number of the sum of Production in 2014?

In [15]:
# Grouping by Element and sum

element_grouping = food_data.groupby('Element').sum()
element_grouping

Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                   708993   14197445      28068795   
Export Quantity                            599910   11840553      26026133   
Fat supply quantity (g/capita/day)         675050   13535000       3435732   
Feed                                       176272    3538507       7282199   
Food                                       663295   13285035      25406622   
Food supply (kcal/capita/day)              674057   13511060       3329296   
Food supply quantity (kg/capita/yr)        658446   13185401       3163725   
Import Quantity                            688174   13795966      28834929   
Losses                                     274353    5424803      10292107   
Other uses (non-food)                      235554    4729749       8926728   
Processing                                 271940    5350416      10313310   
Production                                 526751   10450053      21388191   
Protein supply quantity (g/capita/day)     675050   13535000       3385502   
Residuals                                  623271   12421089      24066350   
Seed                                       103537    2035933       4211574   
Stock Variation                            571566   11329527      21464704   
Total Population - Both sexes                6020     112545         22995   
Tourist consumption                         50308    1486528       2869905   

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein 

In [16]:
# printing the row label with Production

element_grouping.loc['Production']

Area Code         526751.00
Item Code       10450053.00
Element Code    21388191.00
Y2014            1931287.75
Y2015            1947019.39
Y2016            1943537.15
Y2017            2030056.89
Y2018            2075072.89
Name: Production, dtype: float64

    total number of production in 2014
    
            = 1931287.75

# Question 17

    Which of these elements had the highest sum in 2018?

In [17]:
# selecting Year 2018 and Element columns 

element_year = food_data[['Element', 'Y2018']]
element_year

Element     Y2018
0               Total Population - Both sexes  42228.00
1                    Domestic supply quantity      0.00
2               Food supply (kcal/capita/day)   3322.00
3      Protein supply quantity (g/capita/day)     91.83
4          Fat supply quantity (g/capita/day)     77.28
...                                       ...       ...
60938                                    Food     16.00
60939     Food supply quantity (kg/capita/yr)      1.08
60940           Food supply (kcal/capita/day)      1.00
60941  Protein supply quantity (g/capita/day)      0.04
60942      Fat supply quantity (g/capita/day)      0.01

[60943 rows x 2 columns]

In [18]:
element_groupby = element_year.groupby('Element').sum()

element_groupby

Y2018
Element                                           
Domestic supply quantity                2161192.10
Export Quantity                          181594.80
Fat supply quantity (g/capita/day)        10258.69
Feed                                     233489.68
Food                                    1303841.28
Food supply (kcal/capita/day)            455261.00
Food supply quantity (kg/capita/yr)       49056.85
Import Quantity                          287997.09
Losses                                   163902.00
Other uses (non-food)                     91300.97
Processing                               308429.00
Production                              2075072.89
Protein supply quantity (g/capita/day)    11833.56
Residuals                                 34864.00
Seed                                      25263.14
Stock Variation                           20577.91
Total Population - Both sexes           1140605.00
Tourist consumption                          90.00

In [19]:
element_groupby.max()

Y2018    2161192.1
dtype: float64

    the highest sum in 2018
    
            = Domestic supply quantity

# Question 18

    Which of these elements had the 3rd lowest sum in 2018?

In [20]:
# sorting the Year values 

element_groupby.sort_values('Y2018')

Y2018
Element                                           
Tourist consumption                          90.00
Fat supply quantity (g/capita/day)        10258.69
Protein supply quantity (g/capita/day)    11833.56
Stock Variation                           20577.91
Seed                                      25263.14
Residuals                                 34864.00
Food supply quantity (kg/capita/yr)       49056.85
Other uses (non-food)                     91300.97
Losses                                   163902.00
Export Quantity                          181594.80
Feed                                     233489.68
Import Quantity                          287997.09
Processing                               308429.00
Food supply (kcal/capita/day)            455261.00
Total Population - Both sexes           1140605.00
Food                                    1303841.28
Production                              2075072.89
Domestic supply quantity                2161192.10

    the 3rd lowest sum in 2018 
    
            = Protein supply quantity (g/capita/day)

# Question 19

    What is the total Import Quantity in Algeria in 2018?

In [21]:
# sorting Algeria data alone 

algeria = food_data.head(1313)
algeria

Area Code     Area  Item Code           Item  Element Code  \
0             4  Algeria       2501     Population           511   
1             4  Algeria       2501     Population          5301   
2             4  Algeria       2901    Grand Total           664   
3             4  Algeria       2901    Grand Total           674   
4             4  Algeria       2901    Grand Total           684   
...         ...      ...        ...            ...           ...   
1308          4  Algeria       2899  Miscellaneous          5142   
1309          4  Algeria       2899  Miscellaneous           645   
1310          4  Algeria       2899  Miscellaneous           664   
1311          4  Algeria       2899  Miscellaneous           674   
1312          4  Algeria       2899  Miscellaneous           684   

                                     Element             Unit     Y2014  \
0              Total Population - Both sexes     1000 persons  38924.00   
1                   Domestic supply quantity      1000 tonnes      0.00   
2              Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3     Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4         Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                      ...              ...       ...   
1308                                    Food      1000 tonnes     75.00   
1309     Food supply quantity (kg/capita/yr)               kg      1.93   
1310           Food supply (kcal/capita/day)  kcal/capita/day      2.00   
1311  Protein supply quantity (g/capita/day)     g/capita/day      0.06   
1312      Fat supply quantity (g/capita/day)     g/capita/day      0.03   

         Y2015     Y2016     Y2017     Y2018  
0     39728.00  40551.00  41389.00  42228.00  
1         0.00      0.00      0.00      0.00  
2      3379.00   3372.00   3341.00   3322.00  
3        94.35     94.72     92.82     91.83  
4        79.36     77.40     80.19     77.28  
...        ...       ...       ...       ...  
1308     77.00     78.00     80.00     82.00  
1309      1.93      1.93      1.93      1.93  
1310      2.00      2.00      2.00      2.00  
1311      0.06      0.06      0.06      0.06  
1312      0.03      0.03      0.03      0.03  

[1313 rows x 12 columns]

In [22]:
algeria.groupby('Element').sum()

Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                      464     311204        614916   
Export Quantity                               392     264011        579278   
Fat supply quantity (g/capita/day)            432     291489         73872   
Feed                                           88      58967        121462   
Food                                          420     282744        539910   
Food supply (kcal/capita/day)                 432     291489         71712   
Food supply quantity (kg/capita/yr)           420     282744         67725   
Import Quantity                               448     301030        628432   
Losses                                        208     139857        266396   
Other uses (non-food)                         148     101269        190698   
Processing                                    140      93537        179585   
Production                                    360     242590        495990   
Protein supply quantity (g/capita/day)        432     291489         72792   
Residuals                                     408     272327        527340   
Seed                                           80      53109        110540   
Stock Variation                               376     251696        476768   
Total Population - Both sexes                   4       2501           511   

                                           Y2014     Y2015     Y2016  \
Element                                                                
Domestic supply quantity                83755.94  85972.86  87157.61   
Export Quantity                          1328.64   1104.62   1340.22   
Fat supply quantity (g/capita/day)        320.19    317.40    309.54   
Feed                                     9656.04  10308.00   9903.00   
Food                                    60582.69  61693.49  63081.43   
Food supply (kcal/capita/day)           13500.00  13512.00  13483.00   
Food supply quantity (kg/capita/yr)      1556.45   1552.81   1555.77   
Import Quantity                         34940.12  37841.91  36736.08   
Losses                                   7231.00   7471.00   7405.00   
Other uses (non-food)                    3315.22   3483.37   3400.17   
Processing                               1776.00   1865.00   1920.00   
Production                              52534.48  53451.57  54554.75   
Protein supply quantity (g/capita/day)    379.56    377.37    378.84   
Residuals                                 552.00    489.00    770.00   
Seed                                      631.00    657.00    672.00   
Stock Variation                          2392.00   4233.00   2789.00   
Total Population - Both sexes           38924.00  39728.00  40551.00   

                                           Y2017     Y2018  
Element                                                     
Domestic supply quantity                86810.42  90023.42  
Export Quantity                          1494.54    811.54  
Fat supply quantity (g/capita/day)        320.74    309.09  
Feed                                     9571.00  10668.00  
Food                                    62785.33  63736.33  
Food supply (kcal/capita/day)           13363.00  13288.00  
Food supply quantity (kg/capita/yr)      1516.94   1509.28  
Import Quantity                         36059.29  36238.29  
Losses                                   7428.00   7708.00  
Other uses (non-food)                    3534.08   3970.08  
Processing                               2045.00   2479.00  
Production                              54869.67  60766.67  
Protein supply quantity (g/capita/day)    371.26    367.28  
Residuals                                 766.00    630.00  
Seed                                      684.00    825.00  
Stock Variation                          2636.00   6172.00  
Total Population - Both sexes           41389.00  42228.00

    total import Quantity in Algeria in 2018
    
            = 36238.29

# Question 20

    What is the total number of unique countries in the dataset?

In [23]:
food_data['Area'].describe()

count     60943
unique       49
top       Kenya
freq       1560
Name: Area, dtype: object

In [24]:
food_data['Area'].nunique()

49

    total number of unique countries 
    
            = 49